include/k4ainternal/global.h

```diff
@@ -0,0 +1,65 @@
+/** \file global.h
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ * Kinect For Azure SDK.
+ */
+
+#ifndef GLOBAL_H
+#define GLOBAL_H
+
+#include <k4a/k4atypes.h>
+#include <k4ainternal/rwlock.h>
+
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifdef _WIN32
+typedef INIT_ONCE k4a_init_once_t;
+#define K4A_INIT_ONCE INIT_ONCE_STATIC_INIT
+#else
+typedef pthread_once_t k4a_init_once_t;
+#define K4A_INIT_ONCE PTHREAD_ONCE_INIT
+#endif
+
+typedef void(k4a_init_once_function_t)(void);
+
+void global_init_once(k4a_init_once_t *init_once, k4a_init_once_function_t *init_function);
+
+/** Declares an initialized global context
+ *
+ * \param _global_type_
+ * The structure type of the global context
+ *
+ * \param _init_function_
+ * An initialization function that returns void and takes a pointer to a _global_type_ structure.
+ * This function will be called once to initialize the global structure.
+ *
+ * \remarks
+ * This macro creates a new function with the name of the global type followed by _get(). This function
+ * returns a pointer to an initialized singleton instance of the global structure. Initialization will only
+ * happen once per module and is safe across multiple threads attempting to initialize at the same time.
+ *
```

Comment:
I might be missing something, but ```init_function``` will be called every time and needs to be robust to multiple calls.

Reply:
This works. I've added a unit test to verify the behavior.